### PR TITLE
[weight/var_grad] Reduce weight/var_grad dependency for devel

### PR DESCRIPTION
**Pull request by kparichay**
The devel packaging has to be updated, where a lot of headers have to be removed.  This patch makes preparations for it to able to remove weight and var_grad related headers.
- optimizer factory has been deprecated and removed
- parse_utils have been divided into two as it needs exposure for node_exporter
- weight and var_grad are now no longer need for layer_devel requirements

Other minor refactoring is done to make removal of these headers possible.

See Also #986 

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

**Review by zhoonit — COMMENTED** [review]
Good job 👍 it's getting cleaner and nicer, one concern here is that it will break this patch will break some codes in the nntrainer-internal that is using neuralnet headers. Do you have any thoughts for the smooth transition?

**Comment by kparichay** [comment]
> Good job  it's getting cleaner and nicer, one concern here is that it will break this patch will break some codes in the nntrainer-internal that is using neuralnet headers. Do you have any thoughts for the smooth transition?

The update on the nntrainer-internal can be done first, and then merge this patch.

**Comment by kparichay** [comment]
I will separate the last commit for packaging to separate PR. Then nntrainer-internal can be updated first before the actual packaging is updated.

File: nntrainer/layers/layer_context.cpp [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+ * @brief Get regularization loss for the weight
+ *
+ * @param idx Identifier of the weight
+ * @return float Value of the loss
+ */
+float RunLayerContext::getWeightRegularizationLoss(unsigned int idx) const {
+  if (weights[idx]->hasGradient())
+    return weights[idx]->getRegularizationLoss();
+
+  return 0;
```
Comment (zhoonit) [inline nntrainer/layers/layer_context.cpp:52]: I wonder if it is valid to call for regularizationloss when they don't have gradient?
If it is, I guess we don't need check but return `weights[idx]->getRegularizationLoss()`?

File: nntrainer/layers/layer_context.cpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+  return weights[idx]->getGradientRef();
+}
+
+/**
+ * @brief Get regularization loss for the weight
+ *
+ * @param idx Identifier of the weight
+ * @return float Value of the loss
+ */
+float RunLayerContext::getWeightRegularizationLoss(unsigned int idx) const {
```
Comment (zhoonit) [inline nntrainer/layers/layer_context.cpp:48]: maybe check idx and throw?

**Review by zhoonit — APPROVED** [review]
LGTM with minor comments :)

File: nntrainer/layers/layer_context.cpp (hunk 1 above)
Comment (kparichay) [reply nntrainer/layers/layer_context.cpp:52]: A weight can be initialized to be non-trainable but with regularization for now. I guess if that is stopped, then what you said should work.

File: nntrainer/layers/layer_context.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/layers/layer_context.cpp:48]: i will replace `[]` with `.at()` which will do this for us :).

Comment (kparichay) [reply nntrainer/layers/layer_context.cpp:48]: I will update this in a separate PR.

File: nntrainer/layers/layer_context.cpp (hunk 1 above)
Comment (zhoonit) [reply nntrainer/layers/layer_context.cpp]: Cool

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**